<img src="https://wikidocs.net/images/page/22886/rnn_image6between7.PNG" align="left">

# Keras RNN 기초

model.add(SimpleRNN(hidden_units, input_shape=(timesteps, input_dim)))  
model.add(SimpleRNN(hidden_units, input_length=M, input_dim=N))

---

In [1]:
import numpy as np 
import tensorflow as tf
from tensorflow.keras.layers import SimpleRNN, LSTM, Bidirectional

In [2]:
train_X = [[0.1, 4.2, 1.5, 1.1, 2.8], [1.0, 3.1, 2.5, 0.7, 1.1], [0.3, 2.1, 1.5, 2.1, 0.1], [2.2, 1.4, 0.5, 0.9, 1.1]]
print(np.shape(train_X))

(4, 5)


→ 2D 텐서(일반적인 행렬)이기 때문에 3D 텐서로 변환하자

In [3]:
train_X = [[[0.1, 4.2, 1.5, 1.1, 2.8], [1.0, 3.1, 2.5, 0.7, 1.1], [0.3, 2.1, 1.5, 2.1, 0.1], [2.2, 1.4, 0.5, 0.9, 1.1]]]
train_X = np.array(train_X, dtype=np.float32)
print(train_X.shape)

(1, 4, 5)


(batch_size, timesteps, input_dim)

# ---

# SimpleRNN

return_sequences = True → SimpleRNN의 모든 시점의 은닉 상태를 출력함(many-to-many 문제 해결)  
return_sequences = False → SimpleRNN의 마지막 시점만의 은닉 상태를 출력함(many-to-one 문제 해결)

return_state = True → (return_sequences와 관계없이) 마지막 시점의 은닉 상태를 출력함

<img src="https://wikidocs.net/images/page/22886/rnn_image8_ver2.PNG" align="left">

---

1. (return_sequences, return_state) = (False, False)인 경우

In [4]:
rnn = SimpleRNN(3) # SimpleRNN(units, return_sequences=Flase, return_state=False)
hidden_state = rnn(train_X)

print('hidden state : {}, shape : {}'.format(hidden_state, hidden_state.shape))


hidden state : [[ 0.99779236  0.98238194 -0.96625775]], shape : (1, 3)


마지막 시점의 은닉 상태가 출력되는 것을 확인할 수 있음

---

2. (return_sequences, return_state) = (True, False)인 경우

In [5]:
rnn = SimpleRNN(3, return_sequences=True)
hidden_states =  rnn(train_X)

print('hidden states : {}, shape : {}'.format(hidden_states, hidden_states.shape))

hidden states : [[[ 0.97294873 -0.9986201  -0.9852192 ]
  [ 0.7791255   0.23681597 -0.98734236]
  [ 0.8989133  -0.2689766  -0.96348786]
  [ 0.9934039  -0.73291045 -0.8300911 ]]], shape : (1, 4, 3)


모든 시점의 은닉 상태가 출력되는 것을 확인할 수 있음

---

3. (return_sequences, return_state) = (True, True)인 경우

In [6]:
rnn = SimpleRNN(3, return_sequences=True, return_state=True)
hidden_states, last_state = rnn(train_X)

print('hidden states : {}, shape : {}'.format(hidden_states, hidden_states.shape))
print('last hidden state : {}, shape : {}'.format(last_state, last_state.shape))

hidden states : [[[ 0.8806072   0.8279266   0.9968007 ]
  [ 0.22305678 -0.04166216  0.9709169 ]
  [ 0.5725307   0.83857965  0.5900981 ]
  [-0.8363692   0.90119135  0.99560153]]], shape : (1, 4, 3)
last hidden state : [[-0.8363692   0.90119135  0.99560153]], shape : (1, 3)


모든 시점의 은닉 상태와 마지막 은닉 상태를 출력한 것을 확인할 수 있음

---

4. (return_sequences, return_state) = (False, True)인 경우

In [7]:
rnn = SimpleRNN(3, return_sequences=False, return_state=True)
hidden_state, last_state = rnn(train_X)

print('hidden state : {}, shape : {}'.format(hidden_state, hidden_state.shape))
print('last hidden state : {}, shape : {}'.format(last_state, last_state.shape))

hidden state : [[-0.96820366  0.21530943 -0.49535704]], shape : (1, 3)
last hidden state : [[-0.96820366  0.21530943 -0.49535704]], shape : (1, 3)


# ---

# LSTM

1. (return_sequences, return_state) = (False, True)인 경우

→ SimpleRNN과 달리 return_state=True라면 last hidden state와 cell state 둘 다 반환

In [8]:
lstm = LSTM(3, return_sequences=False, return_state=True)
hidden_state, last_state, last_cell_state = lstm(train_X)

print('hidden state : {}, shape : {}'.format(hidden_state, hidden_state.shape))
print('last hidden state : {}, shape : {}'.format(last_state, last_state.shape))
print('last cell state : {}, shape : {}'.format(last_cell_state, last_cell_state.shape))

hidden state : [[ 0.18465963 -0.19729702 -0.27784726]], shape : (1, 3)
last hidden state : [[ 0.18465963 -0.19729702 -0.27784726]], shape : (1, 3)
last cell state : [[ 0.21178982 -0.9648578  -0.55450886]], shape : (1, 3)


---

2. (return_sequences, return_state) = (True, True)인 경우

In [9]:
lstm = LSTM(3, return_sequences=True, return_state=True)
hidden_state, last_state, last_cell_state = lstm(train_X)

print('hidden states : {}, shape : {}'.format(hidden_state, hidden_state.shape))
print('last hidden state : {}, shape : {}'.format(last_state, last_state.shape))
print('last cell state : {}, shape : {}'.format(last_cell_state, last_cell_state.shape))

hidden states : [[[ 0.2603413   0.09060887 -0.1035516 ]
  [ 0.6696156   0.16278873 -0.21831824]
  [ 0.44381034  0.14252222 -0.3680046 ]
  [ 0.7299013   0.40467337 -0.29832104]]], shape : (1, 4, 3)
last hidden state : [[ 0.7299013   0.40467337 -0.29832104]], shape : (1, 3)
last cell state : [[ 1.9762206   0.5365603  -0.47050995]], shape : (1, 3)


# ---

# Bidirectional LSTM

---

출력되는 은닉 상태의 값 고정

In [10]:
k_init = tf.keras.initializers.Constant(value=0.1)
b_init = tf.keras.initializers.Constant(value=0)
r_init = tf.keras.initializers.Constant(value=0.1)

---

1. (return_sequences, return_state) = (False, True)인 경우

In [11]:
bilstm = Bidirectional(LSTM(3, return_sequences=False, return_state=True,
                            kernel_initializer=k_init, bias_initializer=b_init, recurrent_initializer=r_init))
hidden_states, forward_h, forward_c, backward_h, backward_c = bilstm(train_X)

print('hidden state : {}, shape : {}'.format(hidden_states, hidden_states.shape))
print('forward state : {}, shape : {}'.format(forward_h, forward_h.shape))
print('backward state : {}, shape : {}'.format(backward_h, backward_h.shape))

hidden state : [[0.63031393 0.63031393 0.63031393 0.7038734  0.7038734  0.7038734 ]], shape : (1, 6)
forward state : [[0.63031393 0.63031393 0.63031393]], shape : (1, 3)
backward state : [[0.7038734 0.7038734 0.7038734]], shape : (1, 3)


→ hidden state = 정방향 LSTM 마지막 시점의 은닉 상태 + 역방향 LSTM 첫번째 시점의 은닉 상태

→ 그림으로 이해해보자

<img src="https://wikidocs.net/images/page/94748/bilstm3.PNG" align="left">

---

2. (return_sequences, return_state) = (True, True)인 경우

In [12]:
bilstm = Bidirectional(LSTM(3, return_sequences=True, return_state=True,
                           kernel_initializer=k_init, bias_initializer=b_init, recurrent_initializer=r_init))
hidden_states, forward_h, forward_c, backward_h, backward_c = bilstm(train_X)

print('hidden states : {}, shape : {}'.format(hidden_states, hidden_states.shape))
print('forward state : {}, shape : {}'.format(forward_h, forward_h.shape))
print('backward state : {}, shape : {}'.format(backward_h, backward_h.shape))

hidden states : [[[0.35906473 0.35906473 0.35906473 0.7038734  0.7038734  0.7038734 ]
  [0.5511133  0.5511133  0.5511133  0.58863586 0.58863586 0.58863586]
  [0.59115756 0.59115756 0.59115756 0.3951699  0.3951699  0.3951699 ]
  [0.63031393 0.63031393 0.63031393 0.21942244 0.21942244 0.21942244]]], shape : (1, 4, 6)
forward state : [[0.63031393 0.63031393 0.63031393]], shape : (1, 3)
backward state : [[0.7038734 0.7038734 0.7038734]], shape : (1, 3)


→ 그림으로 이해해보자

<img src="https://wikidocs.net/images/page/94748/bilstm1.PNG" align="left">

# ---

# SimpleRNN을 이용한 텍스트 생성(Text Generation using RNN)

1. 전처리

In [13]:
import numpy as np
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.utils import to_categorical

In [14]:
text = """경마장에 있는 말이 뛰고 있다\n
그의 말이 법이다\n
가는 말이 고와야 오는 말이 곱다\n"""

In [15]:
tokenizer = Tokenizer()
tokenizer.fit_on_texts([text]) # text 각각에 index 부여
vocab_size = len(tokenizer.word_index)+1
print('단어 집합의 크기 : %d' %vocab_size)

단어 집합의 크기 : 12


In [16]:
print(tokenizer.word_index)

{'말이': 1, '경마장에': 2, '있는': 3, '뛰고': 4, '있다': 5, '그의': 6, '법이다': 7, '가는': 8, '고와야': 9, '오는': 10, '곱다': 11}


---

(1) 훈련 데이터 생성

In [17]:
sequences = []
for line in text.split('\n'):
    encoded = tokenizer.texts_to_sequences([line])[0]
    for i in range(1, len(encoded)):
        sequence = encoded[:i+1]
        sequences.append(sequence)
        
print('학습에 사용할 샘플의 갯수 : %d' %len(sequences))

학습에 사용할 샘플의 갯수 : 11


In [18]:
print(sequences)

[[2, 3], [2, 3, 1], [2, 3, 1, 4], [2, 3, 1, 4, 5], [6, 1], [6, 1, 7], [8, 1], [8, 1, 9], [8, 1, 9, 10], [8, 1, 9, 10, 1], [8, 1, 9, 10, 1, 11]]


Eg.  
[2, 3] : 경마장에 있는  
[2, 3, 1] : 경마장에 있는 말이

---

(2) 패딩

In [19]:
max_len = max(len(l) for l in sequences)
print('샘플의 최대 길이 : {}'.format(max_len))

샘플의 최대 길이 : 6


In [20]:
sequences = pad_sequences(sequences, maxlen=max_len, padding='pre')
print(sequences)

[[ 0  0  0  0  2  3]
 [ 0  0  0  2  3  1]
 [ 0  0  2  3  1  4]
 [ 0  2  3  1  4  5]
 [ 0  0  0  0  6  1]
 [ 0  0  0  6  1  7]
 [ 0  0  0  0  8  1]
 [ 0  0  0  8  1  9]
 [ 0  0  8  1  9 10]
 [ 0  8  1  9 10  1]
 [ 8  1  9 10  1 11]]


padding 방식을 'pre'로 설정함으로써 maxlen보다 짧은 길이의 샘플의 앞을 0으로 채움

---

(3) 레이블 분리

In [21]:
sequences = np.array(sequences)
X = sequences[:, :-1]
y = sequences[:, -1]

In [22]:
print(X)

[[ 0  0  0  0  2]
 [ 0  0  0  2  3]
 [ 0  0  2  3  1]
 [ 0  2  3  1  4]
 [ 0  0  0  0  6]
 [ 0  0  0  6  1]
 [ 0  0  0  0  8]
 [ 0  0  0  8  1]
 [ 0  0  8  1  9]
 [ 0  8  1  9 10]
 [ 8  1  9 10  1]]


In [23]:
print(y)

[ 3  1  4  5  1  7  1  9 10  1 11]


---

(4) 원-핫 인코딩 수행

In [24]:
y = to_categorical(y, num_classes=vocab_size)
print(y)

[[0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]]


---

2. 모델 설계

In [25]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, Dense, SimpleRNN

In [26]:
embedding_dim = 10
hidden_units = 32

# 모델 생성
model = Sequential()

# 임베딩층
model.add(Embedding(vocab_size, embedding_dim)) # Embedding(input_dim, output_dim)

# 은닉층
model.add(SimpleRNN(hidden_units))

# 출력층
model.add(Dense(vocab_size, activation='softmax'))

# 컴파일
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# 모델 학습
model.fit(X, y, epochs=200, verbose=2)


Epoch 1/200


1/1 - 1s - loss: 2.4819 - accuracy: 0.0909 - 1s/epoch - 1s/step
Epoch 2/200
1/1 - 0s - loss: 2.4653 - accuracy: 0.2727 - 0s/epoch - 0s/step
Epoch 3/200
1/1 - 0s - loss: 2.4489 - accuracy: 0.4545 - 0s/epoch - 0s/step
Epoch 4/200
1/1 - 0s - loss: 2.4325 - accuracy: 0.3636 - 0s/epoch - 0s/step
Epoch 5/200
1/1 - 0s - loss: 2.4159 - accuracy: 0.3636 - 633us/epoch - 633us/step
Epoch 6/200
1/1 - 0s - loss: 2.3991 - accuracy: 0.3636 - 0s/epoch - 0s/step
Epoch 7/200
1/1 - 0s - loss: 2.3820 - accuracy: 0.3636 - 0s/epoch - 0s/step
Epoch 8/200
1/1 - 0s - loss: 2.3645 - accuracy: 0.3636 - 4ms/epoch - 4ms/step
Epoch 9/200
1/1 - 0s - loss: 2.3465 - accuracy: 0.3636 - 4ms/epoch - 4ms/step
Epoch 10/200
1/1 - 0s - loss: 2.3279 - accuracy: 0.3636 - 3ms/epoch - 3ms/step
Epoch 11/200
1/1 - 0s - loss: 2.3087 - accuracy: 0.3636 - 3ms/epoch - 3ms/step
Epoch 12/200
1/1 - 0s - loss: 2.2889 - accuracy: 0.3636 - 4ms/epoch - 4ms/step
Epoch 13/200
1/1 - 0s - loss: 2.2684 - accuracy: 0.3636 - 3ms/epoc

verbose=0 → 출력 정보 X  
verbose=0 → 정보를 상세하게 출력  
verbose=0 → 정보를 함축적으로 출력  

In [27]:
def sentence_generation(model, tokenizer, current_word, n): # 모델, 토크나이저, 현재 단어, 반복할 횟수
    init_word = current_word
    sentence = ''
    
    for _ in range(n):
        # 정수 인코딩 + 패딩
        encoded = tokenizer.texts_to_sequences([current_word])[0]
        encoded = pad_sequences([encoded], maxlen=5, padding='pre')
        
        # X : 현재 단어, y : 예측한 단어
        result = model.predict(encoded, verbose=0)
        result = np.argmax(result, axis=1)
        
        for word, index in tokenizer.word_index.items():
            if index == result:
                break
                
        current_word = current_word + ' ' + word
        
        sentence = sentence + ' ' + word
        
    sentence = init_word + sentence
    
    return sentence

---

실습

In [28]:
print(sentence_generation(model, tokenizer, '경마장에', 4))
print(sentence_generation(model, tokenizer, '그의', 2))
print(sentence_generation(model, tokenizer, '가는', 5))

경마장에 있는 말이 뛰고 있다
그의 말이 법이다
가는 말이 고와야 오는 말이 곱다


# ---

# LSTM

1. 전처리

(1) 데이터 불러오기

In [29]:
import pandas as pd
import numpy as np
from string import punctuation

from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.utils import to_categorical

In [30]:
import warnings
warnings.filterwarnings('ignore')

In [31]:
import os
os.chdir('D:/datasets/')

In [32]:
df = pd.read_csv('ArticlesApril2018.csv')
df.head()

articleID  articleWordCount  \
0  5adf6684068401528a2aa69b               781   
1  5adf653f068401528a2aa697               656   
2  5adf4626068401528a2aa628              2427   
3  5adf40d2068401528a2aa619               626   
4  5adf3d64068401528a2aa60f               815   

                                      byline documentType  \
0                             By JOHN BRANCH      article   
1                           By LISA FRIEDMAN      article   
2                              By PETE WELLS      article   
3  By JULIE HIRSCHFELD DAVIS and PETER BAKER      article   
4             By IAN AUSTEN and DAN BILEFSKY      article   

                                            headline  \
0  Former N.F.L. Cheerleaders’ Settlement Offer: ...   
1  E.P.A. to Unveil a New Rule. Its Effect: Less ...   
2                            The New Noma, Explained   
3                                            Unknown   
4                                            Unknown   

                                            keywords  multimedia     newDesk  \
0  ['Workplace Hazards and Violations', 'Football...          68      Sports   
1  ['Environmental Protection Agency', 'Pruitt, S...          68     Climate   
2  ['Restaurants', 'Noma (Copenhagen, Restaurant)...          66      Dining   
3  ['Macron, Emmanuel (1977- )', 'Trump, Donald J...          68  Washington   
4  ['Toronto, Ontario, Attack (April, 2018)', 'Mu...          68     Foreign   

   printPage              pubDate   sectionName  \
0          0  2018-04-24 17:16:49  Pro Football   
1          0  2018-04-24 17:11:21       Unknown   
2          0  2018-04-24 14:58:44       Unknown   
3          0  2018-04-24 14:35:57        Europe   
4          0  2018-04-24 14:21:21        Canada   

                                             snippet              source  \
0  “I understand that they could meet with us, pa...  The New York Times   
1  The agency plans to publish a new regulation T...  The New York Times   
2  What’s it like to eat at the second incarnatio...  The New York Times   
3  President Trump welcomed President Emmanuel Ma...  The New York Times   
4  Alek Minassian, 25, a resident of Toronto’s Ri...  The New York Times   

  typeOfMaterial                                             webURL  
0           News  https://www.nytimes.com/2018/04/24/sports/foot...  
1           News  https://www.nytimes.com/2018/04/24/climate/epa...  
2           News  https://www.nytimes.com/2018/04/24/dining/noma...  
3           News  https://www.nytimes.com/2018/04/24/world/europ...  
4           News  https://www.nytimes.com/2018/04/24/world/canad...

---

(2) 결측값 확인/제거

In [33]:
print('열의 개수 : ', len(df.columns))
print(df.columns)

열의 개수 :  15
Index(['articleID', 'articleWordCount', 'byline', 'documentType', 'headline',
       'keywords', 'multimedia', 'newDesk', 'printPage', 'pubDate',
       'sectionName', 'snippet', 'source', 'typeOfMaterial', 'webURL'],
      dtype='object')


In [34]:
print(df['headline'].isnull().values.any())

False


headline 열에서 결측값이 없다고 확인됨

In [35]:
headline = []

headline.extend(list(df.headline.values))
headline[:5]

['Former N.F.L. Cheerleaders’ Settlement Offer: $1 and a Meeting With Goodell',
 'E.P.A. to Unveil a New Rule. Its Effect: Less Science in Policymaking.',
 'The New Noma, Explained',
 'Unknown',
 'Unknown']

but, 불필요한 Unknown 값이 있는 걸 확인

<변경 전>

In [36]:
print("총 샘플의 개수 : {}".format(len(headline)))

총 샘플의 개수 : 1324


<제거 후>

In [37]:
headline = [word for word in headline if word != "Unknown"]
print('노이즈값 제거 후 샘플의 개수 : {}'.format(len(headline)))

노이즈값 제거 후 샘플의 개수 : 1214


---

(3) 불필요한 특수 문자 확인/제거

In [38]:
headline[:5]

['Former N.F.L. Cheerleaders’ Settlement Offer: $1 and a Meeting With Goodell',
 'E.P.A. to Unveil a New Rule. Its Effect: Less Science in Policymaking.',
 'The New Noma, Explained',
 'How a Bag of Texas Dirt  Became a Times Tradition',
 'Is School a Place for Self-Expression?']

but, 불필요한 특수문자가 포함된 것을 확인

In [39]:
def repreprocessing(raw_sentence):
    preprocessed_sentence = raw_sentence.encode("utf8").decode("ascii", "ignore")
    # 구두점 제거 → 소문자화
    return ''.join(word for word in preprocessed_sentence if word not in punctuation).lower()

preprocessed_headline = [repreprocessing(x) for x in headline]
preprocessed_headline[:5]

['former nfl cheerleaders settlement offer 1 and a meeting with goodell',
 'epa to unveil a new rule its effect less science in policymaking',
 'the new noma explained',
 'how a bag of texas dirt  became a times tradition',
 'is school a place for selfexpression']

제거

---

(4) sequence 형성

In [40]:
tokenizer = Tokenizer()
tokenizer.fit_on_texts(preprocessed_headline)
vocab_size = len(tokenizer.word_index) + 1
print('단어 집합의 크기 : %d' %vocab_size)

단어 집합의 크기 : 3494


In [41]:
sequences = []

for sentence in preprocessed_headline:
    
    encoded = tokenizer.texts_to_sequences([sentence])[0]
    
    for i in range(1, len(encoded)):
        sequence = encoded[:i+1]
        sequences.append(sequence)
        
sequences[:11]

[[99, 269],
 [99, 269, 371],
 [99, 269, 371, 1115],
 [99, 269, 371, 1115, 582],
 [99, 269, 371, 1115, 582, 52],
 [99, 269, 371, 1115, 582, 52, 7],
 [99, 269, 371, 1115, 582, 52, 7, 2],
 [99, 269, 371, 1115, 582, 52, 7, 2, 372],
 [99, 269, 371, 1115, 582, 52, 7, 2, 372, 10],
 [99, 269, 371, 1115, 582, 52, 7, 2, 372, 10, 1116],
 [100, 3]]

Eg.  
[99, 269] → former nfl  
[99, 269, 371] → former nfl cheerleaders

---

(5) 어떤 정수가 어떤 단어를 살펴보기 위한 작업

In [42]:
index_to_word = {}
for key, value in tokenizer.word_index.items():
    index_to_word[value] = key

print('빈도 수 상의 582번 단어 : {}'.format(index_to_word[582]))

빈도 수 상의 582번 단어 : offer


---

(6) 패딩

패딩하기 위해 최대 길이 알아보기

In [43]:
max_len = max(len(l) for l in sequences)
print('샘플의 최대 길이 : {}'.format(max_len))

샘플의 최대 길이 : 24


In [44]:
sequences = pad_sequences(sequences, maxlen=max_len, padding='pre')
print(sequences[:3])

[[   0    0    0    0    0    0    0    0    0    0    0    0    0    0
     0    0    0    0    0    0    0    0   99  269]
 [   0    0    0    0    0    0    0    0    0    0    0    0    0    0
     0    0    0    0    0    0    0   99  269  371]
 [   0    0    0    0    0    0    0    0    0    0    0    0    0    0
     0    0    0    0    0    0   99  269  371 1115]]


---

(7) 레이블 분리

In [45]:
sequences = np.array(sequences)
X = sequences[:, :-1]
y = sequences[:, -1]

print(X[:3])
print(y[:3])

[[  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0  99]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0  99 269]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0  99 269 371]]
[ 269  371 1115]


In [46]:
y = to_categorical(y, num_classes=vocab_size)
y

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

---

2. 모델 설계

In [47]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, Dense, LSTM

In [48]:
embedding_dim = 10
hidden_units = 128

model = Sequential()
model.add(Embedding(vocab_size, embedding_dim))
model.add(LSTM(hidden_units))
model.add(Dense(vocab_size, activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(X, y, epochs=200, verbose=2)

Epoch 1/200
244/244 - 4s - loss: 7.6411 - accuracy: 0.0282 - 4s/epoch - 18ms/step
Epoch 2/200
244/244 - 2s - loss: 7.1113 - accuracy: 0.0295 - 2s/epoch - 9ms/step
Epoch 3/200
244/244 - 2s - loss: 6.9734 - accuracy: 0.0336 - 2s/epoch - 9ms/step
Epoch 4/200
244/244 - 2s - loss: 6.8519 - accuracy: 0.0400 - 2s/epoch - 9ms/step
Epoch 5/200
244/244 - 2s - loss: 6.7072 - accuracy: 0.0454 - 2s/epoch - 9ms/step
Epoch 6/200
244/244 - 2s - loss: 6.5289 - accuracy: 0.0468 - 2s/epoch - 9ms/step
Epoch 7/200
244/244 - 2s - loss: 6.3198 - accuracy: 0.0518 - 2s/epoch - 9ms/step
Epoch 8/200
244/244 - 2s - loss: 6.1072 - accuracy: 0.0587 - 2s/epoch - 9ms/step
Epoch 9/200
244/244 - 2s - loss: 5.9016 - accuracy: 0.0674 - 2s/epoch - 9ms/step
Epoch 10/200
244/244 - 2s - loss: 5.7050 - accuracy: 0.0686 - 2s/epoch - 9ms/step
Epoch 11/200
244/244 - 2s - loss: 5.5227 - accuracy: 0.0724 - 2s/epoch - 9ms/step
Epoch 12/200
244/244 - 2s - loss: 5.3510 - accuracy: 0.0788 - 2s/epoch - 9ms/step
Epoch 13/200
244/244 - 2

In [49]:
def sentence_generation(model, tokenizer, current_word, n):
    init_word = current_word
    sentence = ''
    
    for _ in range(n):
        encoded = tokenizer.texts_to_sequences([current_word])[0]
        encoded = pad_sequences([encoded], maxlen=max_len-1, padding='pre')
        
        result = model.predict(encoded, verbose=0)
        result = np.argmax(result, axis=1)
        
        for word, index in tokenizer.word_index.items():
            if index == result:
                break
                
        current_word = current_word + ' ' + word
        sentence = sentence + ' ' + word
        
    sentence = init_word + sentence
    return sentence

---

실습

In [50]:
print(sentence_generation(model, tokenizer, 'i', 10))
print(sentence_generation(model, tokenizer, 'how', 10))

i disapprove of school vouchers can i still apply for them
how to make facebook more accountable in turnaround you sorry national


# ---

# Char RNNLM(문자 단위 RNN 언어 모델)

1. 전처리

(1) 파일 불러오기

In [51]:
import warnings
warnings.filterwarnings('ignore')

In [52]:
import numpy as np
import urllib.request
from tensorflow.keras.utils import to_categorical

urllib.request.urlretrieve("http://www.gutenberg.org/files/11/11-0.txt", filename="11-0.txt")

f = open('11-0.txt', 'rb')
sentences = []
for sentence in f:
    sentence = sentence.strip() # 공백 제거
    sentence = sentence.lower() # 소문자화
    sentence = sentence.decode('ascii', 'ignore') # ascii 문자로 디코딩 및 오류 무시
    
    if len(sentence) > 0:
        sentences.append(sentence)
f.close()

잘 됐는지 확인

In [53]:
sentences[:5]

["*** start of the project gutenberg ebook alice's adventures in",
 'wonderland ***',
 '[illustration]',
 'alices adventures in wonderland',
 'by lewis carroll']

In [54]:
total_data = ' '.join(sentences)
print('문자열의 길이 또는 총 문자의 개수 : %d' %len(total_data))

문자열의 길이 또는 총 문자의 개수 : 140323


In [55]:
print(total_data[:200])

*** start of the project gutenberg ebook alice's adventures in wonderland *** [illustration] alices adventures in wonderland by lewis carroll the millennium fulcrum edition 3.0 contents chapter i.    


---

(2) 문자-index, index-문자

In [56]:
char_vocab = sorted(list(set(total_data)))
vocab_size = len(char_vocab)
print('문자 집합의 크기 : {}'.format(vocab_size))

문자 집합의 크기 : 43


26개의 알파벳을 포함하여 총 43개의 문자가 있는 것을 확인

In [57]:
char_to_index = dict((char, index) for index, char in enumerate(char_vocab))
print('문자 집합 : ', char_to_index)

문자 집합 :  {' ': 0, '!': 1, "'": 2, '(': 3, ')': 4, '*': 5, ',': 6, '-': 7, '.': 8, '0': 9, '3': 10, ':': 11, ';': 12, '?': 13, '[': 14, ']': 15, '_': 16, 'a': 17, 'b': 18, 'c': 19, 'd': 20, 'e': 21, 'f': 22, 'g': 23, 'h': 24, 'i': 25, 'j': 26, 'k': 27, 'l': 28, 'm': 29, 'n': 30, 'o': 31, 'p': 32, 'q': 33, 'r': 34, 's': 35, 't': 36, 'u': 37, 'v': 38, 'w': 39, 'x': 40, 'y': 41, 'z': 42}


In [58]:
index_to_char = {}
for key, value in char_to_index.items():
    index_to_char[value] = key

In [59]:
print('index_to_char', index_to_char)

index_to_char {0: ' ', 1: '!', 2: "'", 3: '(', 4: ')', 5: '*', 6: ',', 7: '-', 8: '.', 9: '0', 10: '3', 11: ':', 12: ';', 13: '?', 14: '[', 15: ']', 16: '_', 17: 'a', 18: 'b', 19: 'c', 20: 'd', 21: 'e', 22: 'f', 23: 'g', 24: 'h', 25: 'i', 26: 'j', 27: 'k', 28: 'l', 29: 'm', 30: 'n', 31: 'o', 32: 'p', 33: 'q', 34: 'r', 35: 's', 36: 't', 37: 'u', 38: 'v', 39: 'w', 40: 'x', 41: 'y', 42: 'z'}


---

(3) 훈련 데이터 구성

In [60]:
train_X = 'appl'
train_y = 'pple'

appl을 통해서 pple를 예측하고자 함

In [61]:
seq_length = 60

n_samples = int(np.floor((len(total_data) - 1) / seq_length))
print('만들 수 있는 샘플의 수 : {}'.format(n_samples))

만들 수 있는 샘플의 수 : 2338


In [62]:
train_X = []
train_y = []

for i in range(n_samples):
    X_sample = total_data[i * seq_length : (i+1) * seq_length]
    X_encoded = [char_to_index[c] for c in X_sample]
    train_X.append(X_encoded)

    y_sample = total_data[i * seq_length + 1 : (i+1) * seq_length + 1]
    y_encoded = [char_to_index[c] for c in y_sample]
    train_y.append(y_encoded)

문자를 index로 인코딩해서 sample 구성

In [63]:
print('X 데이터의 첫 번째 샘플 (인코딩) : ', train_X[0])
print('y 데이터의 첫 번째 샘플 (인코딩) : ', train_y[0])
print('-' * 150)
print('X 데이터의 첫 번째 샘플 (디코딩) : ', [index_to_char[i] for i in train_X[0]])
print('y 데이터의 첫 번째 샘플 (디코딩) : ', [index_to_char[i] for i in train_y[0]])

X 데이터의 첫 번째 샘플 (인코딩) :  [5, 5, 5, 0, 35, 36, 17, 34, 36, 0, 31, 22, 0, 36, 24, 21, 0, 32, 34, 31, 26, 21, 19, 36, 0, 23, 37, 36, 21, 30, 18, 21, 34, 23, 0, 21, 18, 31, 31, 27, 0, 17, 28, 25, 19, 21, 2, 35, 0, 17, 20, 38, 21, 30, 36, 37, 34, 21, 35, 0]
y 데이터의 첫 번째 샘플 (인코딩) :  [5, 5, 0, 35, 36, 17, 34, 36, 0, 31, 22, 0, 36, 24, 21, 0, 32, 34, 31, 26, 21, 19, 36, 0, 23, 37, 36, 21, 30, 18, 21, 34, 23, 0, 21, 18, 31, 31, 27, 0, 17, 28, 25, 19, 21, 2, 35, 0, 17, 20, 38, 21, 30, 36, 37, 34, 21, 35, 0, 25]
------------------------------------------------------------------------------------------------------------------------------------------------------
X 데이터의 첫 번째 샘플 (디코딩) :  ['*', '*', '*', ' ', 's', 't', 'a', 'r', 't', ' ', 'o', 'f', ' ', 't', 'h', 'e', ' ', 'p', 'r', 'o', 'j', 'e', 'c', 't', ' ', 'g', 'u', 't', 'e', 'n', 'b', 'e', 'r', 'g', ' ', 'e', 'b', 'o', 'o', 'k', ' ', 'a', 'l', 'i', 'c', 'e', "'", 's', ' ', 'a', 'd', 'v', 'e', 'n', 't', 'u', 'r', 'e', 's', ' ']
y 데이터의 첫 번째 샘플 (디코딩

In [64]:
print(train_X[1])
print('-' * 150)
print(train_y[1])

[25, 30, 0, 39, 31, 30, 20, 21, 34, 28, 17, 30, 20, 0, 5, 5, 5, 0, 14, 25, 28, 28, 37, 35, 36, 34, 17, 36, 25, 31, 30, 15, 0, 17, 28, 25, 19, 21, 35, 0, 17, 20, 38, 21, 30, 36, 37, 34, 21, 35, 0, 25, 30, 0, 39, 31, 30, 20, 21, 34]
------------------------------------------------------------------------------------------------------------------------------------------------------
[30, 0, 39, 31, 30, 20, 21, 34, 28, 17, 30, 20, 0, 5, 5, 5, 0, 14, 25, 28, 28, 37, 35, 36, 34, 17, 36, 25, 31, 30, 15, 0, 17, 28, 25, 19, 21, 35, 0, 17, 20, 38, 21, 30, 36, 37, 34, 21, 35, 0, 25, 30, 0, 39, 31, 30, 20, 21, 34, 28]


X에서 오른쪽으로 한 칸 이동한 것이 y가 되도록 인코딩 완료

---

(4) 원-핫 인코딩

In [65]:
train_X = to_categorical(train_X)
train_y = to_categorical(train_y)

print('train_X의 크기(shape) : {}'.format(train_X.shape))
print('train_y의 크기(shape) : {}'.format(train_y.shape))

train_X의 크기(shape) : (2338, 60, 43)
train_y의 크기(shape) : (2338, 60, 43)


<img src="https://wikidocs.net/images/page/22886/rnn_image6between7.PNG" align="left">

X와 y의 구조를 살펴보면 그림과 같음!

---

2. 모델 설계

TimeDistributed를 이용해서 다대다 구조를 만들 수 있음

In [66]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, LSTM, TimeDistributed

hidden_units = 256

# 모델 생성
model = Sequential()

# LSTM 레이어 쌓기
model.add(LSTM(hidden_units, input_shape=(None, train_X.shape[2]), return_sequences=True))
model.add(LSTM(hidden_units, return_sequences=True))

#
model.add(TimeDistributed(Dense(vocab_size, activation='softmax')))

model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(train_X, train_y, epochs=80, verbose=2)

Epoch 1/80
74/74 - 9s - loss: 3.0266 - accuracy: 0.1851 - 9s/epoch - 124ms/step
Epoch 2/80
74/74 - 5s - loss: 2.6890 - accuracy: 0.2605 - 5s/epoch - 72ms/step
Epoch 3/80
74/74 - 5s - loss: 2.3494 - accuracy: 0.3372 - 5s/epoch - 74ms/step
Epoch 4/80
74/74 - 6s - loss: 2.1910 - accuracy: 0.3752 - 6s/epoch - 77ms/step
Epoch 5/80
74/74 - 6s - loss: 2.0887 - accuracy: 0.4020 - 6s/epoch - 77ms/step
Epoch 6/80
74/74 - 5s - loss: 2.0079 - accuracy: 0.4219 - 5s/epoch - 74ms/step
Epoch 7/80
74/74 - 5s - loss: 1.9350 - accuracy: 0.4411 - 5s/epoch - 74ms/step
Epoch 8/80
74/74 - 5s - loss: 1.8724 - accuracy: 0.4569 - 5s/epoch - 72ms/step
Epoch 9/80
74/74 - 5s - loss: 1.8224 - accuracy: 0.4708 - 5s/epoch - 72ms/step
Epoch 10/80
74/74 - 5s - loss: 1.7732 - accuracy: 0.4829 - 5s/epoch - 72ms/step
Epoch 11/80
74/74 - 5s - loss: 1.7266 - accuracy: 0.4938 - 5s/epoch - 72ms/step
Epoch 12/80
74/74 - 5s - loss: 1.6838 - accuracy: 0.5061 - 5s/epoch - 72ms/step
Epoch 13/80
74/74 - 5s - loss: 1.6411 - accuracy

LTSM layer를 여러 개를 쌓는 이유는 순차 데이터를 더 잘 표현하기 위함 → 복잡한 시간 의존성을 잘 캐치할 수 있음

In [67]:
def sentence_generation(model, length):
    ix = [np.random.randint(vocab_size)]

    y_char = [index_to_char[ix[-1]]]
    print(ix[-1], '번 문자', y_char[-1],'로 예측을 시작!')

    X = np.zeros((1, length, vocab_size)) # 높이, 행, 열

    for i in range(length):
        X[0][i][ix[-1]] = 1
        print(index_to_char[ix[-1]], end="")
        ix = np.argmax(model.predict(X[:, :i+1, :])[0], 1)
        y_char.append(index_to_char[ix[-1]])
    return ('').join(y_char)

In [68]:
result = sentence_generation(model, 100)
print(result)

39 번 문자 w 로 예측을 시작!
1/1 [==============================] - 0s 31ms/step
what latitude or longitude as she could not this moment, the funch whole thingch! thought alice. they


# ---

# Char RNN으로 텍스트 생성하기

1. 데이터에 대한 이해와 전처리

(1) 데이터 가져오기

In [69]:
import numpy as np
from tensorflow.keras.utils import to_categorical

In [70]:
raw_text = '''
I get on with life as a programmer,
I like to contemplate beer.
But when I start to daydream,
My mind turns straight to wine.

Do I love wine more than beer?

I like to use words about beer.
But when I stop my talking,
My mind turns straight to wine.

I hate bugs and errors.
But I just think back to wine,
And I'm happy once again.

I like to hang out with programming and depp learning.
But when left alone,
My mind turns straight to wine.
'''

In [71]:
tokens = raw_text.split() # 공백 기준으로 단어 분리
raw_text = ' '.join(tokens) # 단어들 띄어쓰기로 연결
print(raw_text)

I get on with life as a programmer, I like to contemplate beer. But when I start to daydream, My mind turns straight to wine. Do I love wine more than beer? I like to use words about beer. But when I stop my talking, My mind turns straight to wine. I hate bugs and errors. But I just think back to wine, And I'm happy once again. I like to hang out with programming and depp learning. But when left alone, My mind turns straight to wine.


In [72]:
char_vocab = sorted(list(set(raw_text)))
vocab_size = len(char_vocab)
print('문자 집합 : ', char_vocab)
print('문자 집합의 크기 : ', vocab_size)

문자 집합 :  [' ', "'", ',', '.', '?', 'A', 'B', 'D', 'I', 'M', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'r', 's', 't', 'u', 'v', 'w', 'y']
문자 집합의 크기 :  33


---

(2) 전처리(토큰화)

In [73]:
char_to_index = dict((char, index) for index, char in enumerate(char_vocab))
print(char_to_index)

{' ': 0, "'": 1, ',': 2, '.': 3, '?': 4, 'A': 5, 'B': 6, 'D': 7, 'I': 8, 'M': 9, 'a': 10, 'b': 11, 'c': 12, 'd': 13, 'e': 14, 'f': 15, 'g': 16, 'h': 17, 'i': 18, 'j': 19, 'k': 20, 'l': 21, 'm': 22, 'n': 23, 'o': 24, 'p': 25, 'r': 26, 's': 27, 't': 28, 'u': 29, 'v': 30, 'w': 31, 'y': 32}


---

(3) 샘플 구성

In [74]:
length = 11
sequences = []
for i in range(length, len(raw_text)):
    seq = raw_text[i-length : i]
    sequences.append(seq)
print('총 훈련 샘플의 수 : %d' %len(sequences))

총 훈련 샘플의 수 : 426


---

(4) 정수 인코딩

In [75]:
encoded_sequences = []

for sequence in sequences:
    encoded_sequence = [char_to_index[char] for char in sequence]
    encoded_sequences.append(encoded_sequence)

---

(5) 레이블 분리

In [76]:
encoded_sequences = np.array(encoded_sequences)

X_data = encoded_sequences[:, :-1]
y_data = encoded_sequences[:, -1]

---

(6) 원-핫 인코딩

In [77]:
X_data_one_hot = [to_categorical(encoded, num_classes=vocab_size) for encoded in X_data]
X_data_one_hot = np.array(X_data_one_hot)

y_data_one_hot = to_categorical(y_data, num_classes=vocab_size)

In [78]:
print(X_data_one_hot.shape)

(426, 10, 33)


<img src="https://wikidocs.net/images/page/22886/rnn_image6between7.PNG" align="left">

그림으로 이해하기

In [79]:
# X_data

In [80]:
# print(X_data.shape)

In [81]:
# X_data_one_hot

In [82]:
# print(X_data_one_hot.shape)

---

2. 모델 설계하기

In [83]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, LSTM
from tensorflow.keras.preprocessing.sequence import pad_sequences

In [84]:
hidden_units = 64

model = Sequential() # 모델 생성
model.add(LSTM(hidden_units, input_shape=(X_data_one_hot.shape[1], X_data_one_hot.shape[2]))) # 은닉층
model.add(Dense(vocab_size, activation='softmax')) # 전결합층

model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy']) # 다중클래스 분류

model.fit(X_data_one_hot, y_data_one_hot, epochs=100, verbose=2) # 모델 학습

Epoch 1/100
14/14 - 2s - loss: 3.4674 - accuracy: 0.1221 - 2s/epoch - 151ms/step
Epoch 2/100
14/14 - 0s - loss: 3.3741 - accuracy: 0.1972 - 47ms/epoch - 3ms/step
Epoch 3/100
14/14 - 0s - loss: 3.1487 - accuracy: 0.1972 - 31ms/epoch - 2ms/step
Epoch 4/100
14/14 - 0s - loss: 3.0123 - accuracy: 0.1972 - 31ms/epoch - 2ms/step
Epoch 5/100
14/14 - 0s - loss: 2.9611 - accuracy: 0.1972 - 31ms/epoch - 2ms/step
Epoch 6/100
14/14 - 0s - loss: 2.9407 - accuracy: 0.1972 - 47ms/epoch - 3ms/step
Epoch 7/100
14/14 - 0s - loss: 2.9210 - accuracy: 0.1972 - 31ms/epoch - 2ms/step
Epoch 8/100
14/14 - 0s - loss: 2.9018 - accuracy: 0.1972 - 47ms/epoch - 3ms/step
Epoch 9/100
14/14 - 0s - loss: 2.8785 - accuracy: 0.1972 - 31ms/epoch - 2ms/step
Epoch 10/100
14/14 - 0s - loss: 2.8416 - accuracy: 0.2019 - 47ms/epoch - 3ms/step
Epoch 11/100
14/14 - 0s - loss: 2.8184 - accuracy: 0.1948 - 31ms/epoch - 2ms/step
Epoch 12/100
14/14 - 0s - loss: 2.7651 - accuracy: 0.2066 - 47ms/epoch - 3ms/step
Epoch 13/100
14/14 - 0s -

---

3. 예측

In [85]:
def sentence_generation(model, char_to_index, seq_length, seed_text, n):

    init_text = seed_text
    sentence = ''

    for _ in range(n):
        encoded = [char_to_index[char] for char in seed_text] # 정수 인코딩
        encoded = pad_sequences([encoded], maxlen=seq_length, padding='pre') # 패딩
        encoded = to_categorical(encoded, num_classes=len(char_to_index)) # 원-핫

        result = model.predict(encoded, verbose=0) 
        result = np.argmax(result, axis=1)

        for char, index in char_to_index.items():
            if index == result:
                break

        seed_text = seed_text + char

        sentence = sentence + char

    sentence = init_text + sentence
    return sentence

In [86]:
print(sentence_generation(model, char_to_index, 10 , 'I get on w', 80))

I get on with laoerr, s mik ttooann, Mymiind turn sstraight to wine. I hate bugs and error
